checker/cost.go

```diff
@@ -0,0 +1,536 @@
+type Expr interface {
+	// Type returns the deduced type of the Expr.
+	Type() *exprpb.Type
+	// LiteralSize returns the size of the Expr if it is a literal defined inline in CEL, or
+	// nil if the Exprs size is not statically known.
+	LiteralSize() *int64
+}
+
+type expr struct {
+	t    *exprpb.Type
+	expr *exprpb.Expr
+}
+
+func (e expr) Type() *exprpb.Type {
+	return e.t
+}
+
+func (e expr) LiteralSize() *int64 {
+	var v int64
+	switch ek := e.expr.ExprKind.(type) {
+	case *exprpb.Expr_ConstExpr:
+		switch ck := ek.ConstExpr.ConstantKind.(type) {
+		case *exprpb.Constant_StringValue:
+			v = int64(len(ck.StringValue))
+		case *exprpb.Constant_BytesValue:
+			v = int64(len(ck.BytesValue))
+		default:
+			return nil
+		}
+	case *exprpb.Expr_ListExpr:
+		v = int64(len(ek.ListExpr.Elements))
+	case *exprpb.Expr_StructExpr:
+		if ek.StructExpr.MessageName == "" {
+			v = int64(len(ek.StructExpr.Entries))
+		}
+	default:
+		return nil
+	}
+
+	return &v
+}
+
+var (
+	identCost  = CostEstimate{Min: 1, Max: 1}
+	selectCost = CostEstimate{Min: 1, Max: 1}
+	constCost  = CostEstimate{Min: 0, Max: 0}
+)
+
+// CostEstimate represents an estimated cost range.
+type CostEstimate struct {
+	Min, Max int64
+}
+
+// Add adds the costs and returns the sum.
+func (ce CostEstimate) Add(rhs CostEstimate) CostEstimate {
+	return CostEstimate{
+		ce.Min + rhs.Min,
+		ce.Max + rhs.Max,
+	}
+}
+
+type coster struct {
+	env                *Env
+	mappings           *mapping
+	freeTypeVarCounter int
+	sourceInfo         *exprpb.SourceInfo
+	types              map[int64]*exprpb.Type
+	references         map[int64]*exprpb.Reference
+	estimator          CostEstimator
+}
+
+// Cost estimates the cost of the parsed CEL expression.
+// The parsedExpr is expected to have passed type checking. If it has not been
+// type checked, the results are undefined.
+func Cost(parsedExpr *exprpb.ParsedExpr,
+	env *Env, estimator CostEstimator) CostEstimate {
+	c := coster{
+		env:        env,
+		mappings:   newMapping(),
+		sourceInfo: parsedExpr.GetSourceInfo(),
+		types:      make(map[int64]*exprpb.Type),
+		references: make(map[int64]*exprpb.Reference),
+		estimator:  estimator,
+	}
+	return c.cost(parsedExpr.GetExpr())
+}
+
+func (c *coster) cost(e *exprpb.Expr) CostEstimate {
+	if e == nil {
+		return CostEstimate{}
+	}
+
+	var cost CostEstimate
+	switch e.ExprKind.(type) {
+	case *exprpb.Expr_ConstExpr:
+		cost = constCost
+	case *exprpb.Expr_IdentExpr:
+		cost = c.costIdent(e)
+	case *exprpb.Expr_SelectExpr:
+		cost = c.costSelect(e)
+	case *exprpb.Expr_CallExpr:
+		cost = c.costCall(e)
+	case *exprpb.Expr_ListExpr:
+		cost = c.costCreateList(e)
+	case *exprpb.Expr_StructExpr:
+		cost = c.costCreateStruct(e)
+	case *exprpb.Expr_ComprehensionExpr:
+		cost = c.costComprehension(e)
+	default:
+		return CostEstimate{}
+	}
+	return cost
+}
+
+func (c *coster) costIdent(e *exprpb.Expr) CostEstimate {
+	identExpr := e.GetIdentExpr()
```

Comment:
Ident would be the top-level variable. Usually this will be an `identCost`. There should be no need to iterate through the identifiers in the environment since the type-checker will have already done this for you and rewritten the expression to ensure that only resolved identifiers appear in `IdentExpr` nodes.

Reply:
I agree. This should be removed. I hit a snag and added a TODO to try to explain. It might be worth syncing offline about this one. I'm not sure I understand the codebase well enough to resolve the problem.

Reply:
I added a comment about how to use the `CheckedExpr.type_map` near the `coster` struct definition. Hopefully that helps with removing the need for the `c.env.IdentLookup`

Reply:
https://github.com/google/cel-go/pull/483/commits/49c0f67e25e528a1d7827caf6cd00d972e98de28#r809444558